src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/ProtobufSerializerHelper.cs

```diff
@@ -0,0 +1,233 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Buffers.Binary;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace OpenTelemetry.Exporter.Geneva;
+
+internal static class ProtobufSerializerHelper
+{
+    private const int Fixed64Size = 8;
+
+    internal static Encoding Utf8Encoding => Encoding.UTF8;
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static void WriteStringTag(byte[] buffer, ref int cursor, int fieldNumber, string value)
+    {
+        int stringSize = Utf8Encoding.GetByteCount(value);
+
+        WriteTag(buffer, ref cursor, fieldNumber, WireFormat.WireType.LengthDelimited);
+
+        WriteLength(buffer, ref cursor, stringSize);
+
+        _ = Encoding.UTF8.GetBytes(value, 0, value.Length, buffer, cursor);
+
+        cursor += stringSize;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static void WriteEnumWithTag(byte[] buffer, ref int cursor, int fieldNumber, int value)
+    {
+        WriteTag(buffer, ref cursor, fieldNumber, WireFormat.WireType.Varint);
+
+        // Assuming 1 byte which matches the intended use.
+        // Otherwise, need to first calculate the bytes needed.
+        WriteRawByte(buffer, ref cursor, (byte)value);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static void WriteBoolWithTag(byte[] buffer, ref int cursor, int fieldNumber, bool value)
+    {
+        WriteTag(buffer, ref cursor, fieldNumber, WireFormat.WireType.Varint);
+        WriteRawByte(buffer, ref cursor, value ? (byte)1 : (byte)0);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static void WriteFixed64WithTag(byte[] buffer, ref int cursor, int fieldNumber, ulong value)
+    {
+        WriteTag(buffer, ref cursor, fieldNumber, WireFormat.WireType.Fixed64);
+        WriteRawLittleEndian64(buffer, ref cursor, value);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static void WriteInt64WithTag(byte[] buffer, ref int cursor, int fieldNumber, ulong value)
+    {
+        WriteTag(buffer, ref cursor, fieldNumber, WireFormat.WireType.Varint);
+        WriteRawVarint64(buffer, ref cursor, value);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static void WriteDoubleWithTag(byte[] buffer, ref int cursor, int fieldNumber, double value)
+    {
+        WriteTag(buffer, ref cursor, fieldNumber, WireFormat.WireType.Fixed64);
+        WriteRawLittleEndian64(buffer, ref cursor, (ulong)BitConverter.DoubleToInt64Bits(value));
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static void WriteLengthCustom(byte[] buffer, ref int cursor, int length)
+    {
+        WriteRawVarintCustom(buffer, ref cursor, (uint)length);
```

Comment:
Should we avoid calling WriteRawVarintCustom and have the implementation here, as this is custom?